include/openmc/source.h

```diff
@@ -59,14 +74,45 @@ class SourceDistribution {
   UPtrDist energy_; //!< Energy distribution
 };
 
-class CustomSource {
-  public:
-    virtual ~CustomSource() {}
+//==============================================================================
+//! Source composed of particles read from a file
+//==============================================================================
+
+class FileSource : public Source {
+public:
+  // Constructors
+  explicit FileSource(std::string path);
+
+  // Methods
+  Particle::Bank sample(uint64_t* seed) override;
+
+private:
+  std::vector<Particle::Bank> sites_; //!< Source sites from a file
+};
+
+//==============================================================================
+//! Wrapper for custom sources that manages opening/closing shared library
+//==============================================================================
+
+class CustomSourceWrapper : public Source {
+public:
+  // Constructors, destructors
+  CustomSourceWrapper(std::string path, std::string parameters);
+  ~CustomSourceWrapper();
+
+  // Defer implementation to custom source library
+  Particle::Bank sample(uint64_t* seed) override
+  {
+    return custom_source_->sample(seed);
+  }
 
-    virtual Particle::Bank sample(uint64_t* seed) = 0;
+  double strength() const override { return custom_source_->strength(); }
```

Comment:
just checking, why does this need a concrete function?

Reply:
Yes, it will inherit from `Source`, but if the person writing the custom source has implemented `strength()`, the wrapper needs to use that version.